HeterogeneousCore/CUDACore/interface/CUDAContextState.h

```diff
@@ -0,0 +1,57 @@
+#ifndef HeterogeneousCore_CUDACore_CUDAContextState_h
+#define HeterogeneousCore_CUDACore_CUDAContextState_h
+
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h"
+
+#include <memory>
+
+/**
+ * The purpose of this class is to deliver the device and CUDA stream
+ * information from ExternalWork's acquire() to producer() via a
+ * member/StreamCache variable.
+ */
+class CUDAContextState {
+public:
+  CUDAContextState() = default;
+  ~CUDAContextState() = default;
+
+  CUDAContextState(const CUDAContextState&) = delete;
+  CUDAContextState& operator=(const CUDAContextState&) = delete;
+  CUDAContextState(CUDAContextState&&) = delete;
+  CUDAContextState& operator=(CUDAContextState&& other) = delete;
+
+private:
+  friend class CUDAScopedContextAcquire;
+  friend class CUDAScopedContextProduce;
+  friend class CUDAScopedContextTask;
+
+  void set(int device, cudautils::SharedStreamPtr stream) {
```

Comment:
Would a `using` be helpful for the type used to denote the device?

Reply:
> Would a `using` be helpful for the type used to denote the device?

You mean something like `using DeviceType = int`? The CUDA runtime API uses `int`, and I'm a bit afraid that adding our own type in between (even if just `using`) would create confusion.